src/background/utils/requests.js

```diff
@@ -107,52 +108,61 @@ const HeaderInjector = (() => {
   };
 })();
 
-export function getRequestId() {
+export function getRequestId(eventsToNotify = []) {
+  eventsToNotify.push('loadend');
   const id = getUniqId();
   requests[id] = {
     id,
+    eventsToNotify,
     xhr: new XMLHttpRequest(),
   };
   return id;
 }
 
 function xhrCallbackWrapper(req) {
-  let lastPromise = Promise.resolve();
-  const { xhr } = req;
-  return (evt) => {
-    const res = {
-      id: req.id,
-      type: evt.type,
-      resType: xhr.responseType,
-      contentType: xhr.getResponseHeader('Content-Type') || 'application/octet-stream',
-    };
+  let lastPromise;
+  let blobUrl;
+  const { id, xhr } = req;
+  return async (evt) => {
+    if (evt.type === 'loadend') clearRequest(req);
+    if (!req.cb) return;
     const data = {
       finalUrl: xhr.responseURL,
       readyState: xhr.readyState,
       responseHeaders: req.responseHeaders || xhr.getAllResponseHeaders(),
       status: xhr.status,
       statusText: xhr.statusText,
+      response: xhr.responseType === 'blob' && xhr.response ? getBlobUrl(req) : xhr.response,
+      ...tryOrIgnore(objectPick, xhr, ['responseText']),
+      ...('loaded' in evt) && objectPick(evt, ['lengthComputable', 'loaded', 'total']),
     };
-    res.data = data;
+    await lastPromise;
+    lastPromise = req.cb({
+      id,
+      data,
+      contentType: xhr.getResponseHeader('Content-Type') || 'application/octet-stream',
+      resType: xhr.responseType,
+      type: evt.type,
+    });
+  };
+  function getBlobUrl() {
+    if (!blobUrl) {
+      blobUrl = URL.createObjectURL(xhr.response);
+      setTimeout(expireBlobUrl, 60e3);
+    }
+    return blobUrl;
+  }
+  function expireBlobUrl() {
+    URL.revokeObjectURL(blobUrl);
+    blobUrl = null;
+  }
+  function tryOrIgnore(fn, ...args) {
```

Comment:
`tryOrIgnore` does not depend on any scoped variables so it's better to move it to the top level.

Personally I prefer a higher order function like this:

```js
function ignoreError(fn) {
  const wrapped = (...args) => {
    try {
      return fn(...args);
    } catch {
      // ignore
    }
  };
  return wrapped;
}
```

so that we don't need to change the way we call `fn`:
```js
fn(a, b);
ignoreError(fn)(a, b);

// as a drop-in replacement
fn = ignoreError(fn);
fn(a, b);
```

Reply:
That's a leftover from my earlier approach when I thought it'd be helpful to use this (or similar) function in more places but I found only one or two where it would improve readability. You're right so now I've reverted to a simple scoped function.